### PR TITLE
Add sql on fhir library

### PR DESCRIPTION
**Pull request by mahimadesilva**
## Purpose
Add SQL on FHIR library.

## Summary

This pull request introduces a comprehensive SQL on FHIR library (`sqlonfhir`) to the Ballerina FHIR module. The library enables evaluation and transpilation of FHIR ViewDefinition queries against FHIR R4 resources.

## Key Features

**In-Memory Runner Module** (`sql_on_fhir_lib.in_memory_runner`)
- Evaluates ViewDefinition resources against FHIR JSON data in memory
- Supports FHIRPath expression evaluation with injectable extension functions
- Implements column selection, filtering via `where` clauses, and collection operations
- Handles `forEach`, `forEachOrNull`, and `repeat` constructs for data traversal
- Supports `unionAll` operations for combining multiple query branches

**PostgreSQL Runner Module** (`sql_on_fhir_lib.pg_db_runner`)
- Transpiles ViewDefinition queries to PostgreSQL SQL statements
- Includes a complete FHIRPath expression parser and scanner
- Generates `SELECT` statements with support for recursive CTEs, lateral joins, and window operations
- Supports complex query construction including `forEach`, `repeat`, and `unionAll` operations in SQL

## Implementation Details

- **Core Expression Support**: FHIRPath operators (arithmetic, comparison, boolean logic), function calls, collection projections, type filtering (`ofType`), and string operations
- **Constant Management**: Support for view-level constants with typed values (base64Binary, date, decimal, etc.)
- **Array Operations**: Indexing, first/last element selection, existence checks, and join operations
- **Extension Support**: Custom FHIRPath extension functions with default implementations for resource and reference key resolution
- **Test Coverage**: Over 100 test cases covering in-memory evaluation scenarios and PostgreSQL query generation, including edge cases for collections, nested structures, and complex filtering

## CI/CD Updates

- New manual deployment workflow (`cd-sqlonfhir.yml`) with PostgreSQL service provisioning
- Updated CI workflow to run dedicated in-memory tests for the sqlonfhir module
- New daily build workflow (`daily-build-sqlonfhir.yml`) with PostgreSQL support and coverage reporting

## Manifest Files

- Created `Ballerina.toml` and `Dependencies.toml` defining the package structure, exports, and dependency management
- Added comprehensive README documenting library capabilities and usage examples

### CODE REVIEW

File: sqlonfhir/modules/pg_db_runner/tests/fn_boundary_pg_test.bal [hunk 1]
```diff
@@ -0,0 +1,554 @@
...
+    stream<record {}, sql:Error?> resultStream = dbClient->query(`SELECT * FROM sof_test_view`);
+    json[] result = [];
+    check from record {} row in resultStream
+        do {
+            result.push(row.toJson());
+        };
+    _ = check dbClient->execute(`DROP VIEW IF EXISTS sof_test_view`);
+    assertResultsMatch(result, expected);
+    check dbClient.close();
+}
```
[the remark this answers is not in the document]
Comment (mahimadesilva) [reply sqlonfhir/modules/pg_db_runner/tests/fn_boundary_pg_test.bal:569]: These are disabled due to current limitaions of the fhirpath library, will be enabled when those are implemented

File: sqlonfhir/modules/pg_db_runner/scanner.bal [hunk 2]
```diff
@@ -0,0 +1,415 @@
...
+# Returns a null character if beyond the end of the source.
+#
+# + state - The current scanner state
+# + return - The next character, or "\0" if at or past end of source
+isolated function peekScannerNext(ScannerState state) returns string {
+    if state.current + 1 >= state.sourceCode.length() {
+        return "\\0";
+    }
+    return charAt(state.sourceCode, state.current + 1);
+}
```
[the remark this answers is not in the document]
Comment (mahimadesilva) [reply sqlonfhir/modules/pg_db_runner/scanner.bal:317]: This is not an issue

[the remark this answers is not in the document]
Comment (mahimadesilva) [reply sqlonfhir/modules/pg_db_runner/scanner.bal:317]: there is no sentinel null character in bal

File: sqlonfhir/modules/pg_db_runner/tests/combinations_pg_test.bal [hunk 3]
```diff
@@ -0,0 +1,400 @@
...
+    stream<record {}, sql:Error?> resultStream = dbClient->query(`SELECT * FROM sof_test_view`);
+    json[] result = [];
+    check from record {} row in resultStream
+        do {
+            result.push(row.toJson());
+        };
+    _ = check dbClient->execute(`DROP VIEW IF EXISTS sof_test_view`);
+    assertResultsMatch(result, expected);
+    check dbClient.close();
+}
```
[the remark this answers is not in the document]
Comment (mahimadesilva) [reply sqlonfhir/modules/pg_db_runner/tests/combinations_pg_test.bal:414]: This is true. Theses tests were generated from athe standard spec test suite. That test suite has it incorrectly.

File: sqlonfhir/modules/pg_db_runner/parser.bal [hunk 4]
```diff
@@ -0,0 +1,522 @@
...
+            return [createFunctionExpr(name, (), params), newState];
+        }
+
+        // Just an identifier
+        return [createIdentifierExpr(name), newState];
+    }
+
+    FhirPathToken token = peekparse(state);
+    return error FhirpathParserError("Expect expression.",
+        token = token);
```
[the remark this answers is not in the document]
Comment (mahimadesilva) [reply sqlonfhir/modules/pg_db_runner/parser.bal:432]: https://github.com/wso2-enterprise/open-healthcare/issues/2120

File: sqlonfhir/modules/in_memory_runner/tests/constant_test.bal [hunk 5]
```diff
@@ -0,0 +1,353 @@
...
+                        path: "family",
+                        'type: "string"
+                    }
+                ]
+            }
+        ]
+    };
+    json[]|error result = evaluate(constantResources, view);
+    test:assertTrue(result is error, msg = "Expected an error for: accessing an undefined constant");
+}
```
[the remark this answers is not in the document]
Comment (mahimadesilva) [reply sqlonfhir/modules/in_memory_runner/tests/constant_test.bal:322]: not a major issue

File: sqlonfhir/modules/in_memory_runner/tests/validate_test.bal [hunk 6]
```diff
@@ -0,0 +1,62 @@
...
+        ],
+        'where: [
+            {
+                path: "name.family"
+            }
+        ]
+    };
+    json[]|error result = evaluate(validateResources, view);
+    test:assertTrue(result is error, msg = "Expected an error for: where with path resolving to not boolean");
+}
```
[the remark this answers is not in the document]
Comment (mahimadesilva) [reply sqlonfhir/modules/in_memory_runner/tests/validate_test.bal:77]: not a major issue

**Review by sameeragunarathne — APPROVED** [review]
